pkg/cloudprovider/vsphere/nodemanager_test.go

```diff
@@ -0,0 +1,227 @@
+// The vcsim instance is configured with a tls.Config. The returned client
+// config can be configured to allow/decline insecure connections.
+func configFromSimWithTLS(tlsConfig *tls.Config, insecureAllowed bool) (Config, func()) {
+	var cfg Config
+	model := simulator.VPX()
+
+	err := model.Create()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	model.Service.TLS = tlsConfig
+	s := model.Service.NewServer()
+
+	// STS simulator
+	path, handler := sts.New(s.URL, vpx.Setting)
+	model.Service.ServeMux.Handle(path, handler)
+
+	// Lookup Service simulator
+	model.Service.RegisterSDK(lookup.New())
+
+	cfg.Global.InsecureFlag = insecureAllowed
+
+	cfg.Global.VCenterIP = s.URL.Hostname()
+	cfg.Global.VCenterPort = s.URL.Port()
+	cfg.Global.User = s.URL.User.Username()
+	cfg.Global.Password, _ = s.URL.User.Password()
+	cfg.Global.Datacenters = vclib.TestDefaultDatacenter
+	cfg.Network.PublicNetwork = vclib.TestDefaultNetwork
+	cfg.VirtualCenter = make(map[string]*VirtualCenterConfig)
+	cfg.VirtualCenter[s.URL.Hostname()] = &VirtualCenterConfig{}
+
+	return cfg, func() {
+		s.Close()
+		model.Remove()
+	}
+}
+
+// configFromEnvOrSim returns config from configFromEnv if set, otherwise returns configFromSim.
+func configFromEnvOrSim() (Config, func()) {
+	cfg, ok := configFromEnv()
+	if ok {
+		return cfg, func() {}
+	}
+	return configFromSim()
+}
+
+func TestRegUnregNode(t *testing.T) {
+	cfg, ok := configFromEnvOrSim()
+	defer ok()
+
+	vsphere, err := newVSphere(cfg)
+	if err != nil {
+		t.Fatalf("Failed to construct/authenticate vSphere: %s", err)
+	}
+
+	nm := NodeManager{
+		vsphereInstanceMap: vsphere.vsphereInstanceMap,
+		nodeNameMap:        make(map[string]*NodeInfo),
+		nodeUUIDMap:        make(map[string]*NodeInfo),
+		nodeRegUUIDMap:     make(map[string]*v1.Node),
+	}
+
+	name := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine).Config.GuestFullName
+	UUID := simulator.Map.Any("VirtualMachine").(*simulator.VirtualMachine).Config.Uuid
+	k8sUUID := nm.convertK8sUUIDtoNormal(UUID)
+
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Status: v1.NodeStatus{
+			NodeInfo: v1.NodeSystemInfo{
+				SystemUUID: k8sUUID,
+			},
+		},
+	}
+
+	nm.RegisterNode(node)
+
+	if len(nm.nodeNameMap) != 1 {
+		t.Errorf("Failed: nodeNameMap should be a length of 1")
+	}
+	if len(nm.nodeUUIDMap) != 1 {
+		t.Errorf("Failed: nodeUUIDMap should be a length of  1")
+	}
+	if len(nm.nodeRegUUIDMap) != 1 {
+		t.Errorf("Failed: nodeRegUUIDMap should be a length of  1")
+	}
+
+	nm.UnregisterNode(node)
+
+	if len(nm.nodeNameMap) != 1 {
+		t.Errorf("Failed: nodeNameMap should be a length of  1")
+	}
+	if len(nm.nodeUUIDMap) != 1 {
+		t.Errorf("Failed: nodeUUIDMap should be a length of  1")
+	}
+	if len(nm.nodeRegUUIDMap) != 0 {
+		t.Errorf("Failed: nodeRegUUIDMap should be a length of 0")
+	}
+}
+
+//Sim error: SearchIndex:SearchIndex does not implement: FindByDnsName
```

Comment:
We can implement that method in vcsim.  Can you open an issue in the govmomi repo?

Reply:
Will do!

Reply:
https://github.com/vmware/govmomi/issues/1231